CGR HTTP client double encodes URL parameters

## What happened?

<!--
Describe what you've observed and why it's bad (please include whatever stacktraces/version numbers you can).
Clear steps to reproduce the behaviour are always helpful too!
-->
The CGR HTTP client will incorrectly double encode URL parameters, which was introduced by the following change: https://github.com/palantir/conjure-go-runtime/pull/155

Conjure generated clients already encode URL parameters a single time when building the request, so every param will already be encoded when the `client.Do()` method is called. However, the `client.Do()` will construct the URI again and when building it will re-encode one more time, making the original request incorrect.
For example, the following shows the initial client URL parameter encoding when building the request path:
https://github.com/palantir/conjure-go/blob/aaad4d3e22ebf4e48816fb1597fb7153dde97895/integration_test/testgenerated/client/api/services.conjure.go#L91
Then in the `client.doOnce` function, it will call out to `joinURIandPath` ([link](https://github.com/palantir/conjure-go-runtime/blob/929262581a760b34308c018681c8f63466d17ee6/conjure-go-client/httpclient/client.go#L226-L237)), which builds the string and encodes parameters one more time.
The following test will repro the issue
```go
func TestDoubleURLEncoding(t *testing.T) {
	actual, err := joinURIAndPath("https://localhost", "/api/+t%2FA=")
	require.NoError(t, err)
	assert.Equal(t, "https://localhost/api/+t%2FA=", actual)
}
```
Where the result is the following:
```
=== RUN   TestDoubleURLEncoding
    client_path_test.go:27: 
        	Error Trace:	client_path_test.go:27
        	Error:      	Not equal: 
        	            	expected: "https://localhost/api/+t%2FA="
        	            	actual  : "https://localhost/api/+t%252FA="
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1 +1 @@
        	            	-https://localhost/api/+t%2FA=
        	            	+https://localhost/api/+t%252FA=
        	Test:       	TestDoubleURLEncoding
--- FAIL: TestDoubleURLEncoding (0.00s)
FAIL
FAIL	github.com/palantir/conjure-go-runtime/v2/conjure-go-client/httpclient	0.212s
FAIL
```
As you can see there is a double encoding which results in a `%252F` for the original `%2F` which is the encoded form of a `/`

## What did you want to happen?

The encoded URL params should be preserved and not encoded again.
In the concrete example, the URL param is the following:
```
+t/A=
```
The `/` is encoded as `%2F` when the request is built and thus the full request path should be
```
https://localhost/api/+t%2FA=
```

<!--
Suggest better behaviour
-->

--- a/changelog/@unreleased/pr-182.v2.yml
+++ b/changelog/@unreleased/pr-182.v2.yml
@@ -1,0 +1,7 @@
+type: fix
+fix:
+  description: |-
+    Avoid double encoding URL parameters
+    Fixes: #181
+  links:
+  - https://github.com/palantir/conjure-go-runtime/pull/182

--- a/conjure-go-client/httpclient/client.go
+++ b/conjure-go-client/httpclient/client.go
@@ -137,10 +137,7 @@ func (c *clientImpl) doOnce(
 	if b.method == "" {
 		return nil, werror.Error("httpclient: use WithRequestMethod() to specify HTTP method")
 	}
-	reqURI, err := joinURIAndPath(baseURI, b.path)
-	if err != nil {
-		return nil, err
-	}
+	reqURI := joinURIAndPath(baseURI, b.path)
 	req, err := http.NewRequest(b.method, reqURI, nil)
 	if err != nil {
 		return nil, werror.Wrap(err, "failed to build new HTTP request")
@@ -223,15 +220,10 @@ func (c *clientImpl) initializeRequestHeaders(ctx context.Context) http.Header {
 	return headers
 }
 
-func joinURIAndPath(baseURI, reqPath string) (string, error) {
-	uri, err := url.Parse(baseURI)
-	if err != nil {
-		return "", werror.Wrap(err, "failed to parse request URL")
-	}
-
+func joinURIAndPath(baseURI, reqPath string) string {
+	fullURI := strings.Trim(baseURI, "/")
 	if reqPath != "" {
-		uri.Path = strings.TrimRight(uri.Path, "/") + "/" + strings.TrimLeft(reqPath, "/")
+		fullURI += "/" + strings.TrimLeft(reqPath, "/")
 	}
-
-	return uri.String(), nil
+	return fullURI
 }

--- a/conjure-go-client/httpclient/client.go
+++ b/conjure-go-client/httpclient/client.go
@@ -221,7 +221,7 @@ func (c *clientImpl) initializeRequestHeaders(ctx context.Context) http.Header {
 }
 
 func joinURIAndPath(baseURI, reqPath string) string {
-	fullURI := strings.Trim(baseURI, "/")
+	fullURI := strings.TrimRight(baseURI, "/")
 	if reqPath != "" {
 		fullURI += "/" + strings.TrimLeft(reqPath, "/")
 	}

--- a/conjure-go-client/httpclient/client_path_test.go
+++ b/conjure-go-client/httpclient/client_path_test.go
@@ -18,7 +18,6 @@ import (
 	"testing"
 
 	"github.com/stretchr/testify/assert"
-	"github.com/stretchr/testify/require"
 )
 
 func TestJoinURIandPath(t *testing.T) {
@@ -72,10 +71,14 @@ func TestJoinURIandPath(t *testing.T) {
 			"",
 			"https://localhost/api",
 		},
+		{
+			"https://localhost",
+			"/api/+ti%2FQojjmKJxpxmY%2FA=",
+			"https://localhost/api/+ti%2FQojjmKJxpxmY%2FA=",
+		},
 	} {
 		t.Run("", func(t *testing.T) {
-			actual, err := joinURIAndPath(test.baseURI, test.reqPath)
-			require.NoError(t, err)
+			actual := joinURIAndPath(test.baseURI, test.reqPath)
 			assert.Equal(t, test.expected, actual)
 		})
 	}

